src/coreclr/jit/promotion.cpp

```diff
@@ -0,0 +1,1309 @@
+            countOverlappedCallsWtd + countOverlappedReturnsWtd + countOverlappedAssignmentSourceWtd;
+        costWith += countWriteBacksWtd * writeBackCost;
+
+        JITDUMP("Evaluating access %s @ %03u\n", varTypeName(access.AccessType), access.Offset);
+        JITDUMP("  Single write-back cost: " FMT_WT "\n", writeBackCost);
+        JITDUMP("  Write backs: " FMT_WT "\n", countWriteBacksWtd);
+        JITDUMP("  Read backs: " FMT_WT "\n", countReadBacksWtd);
+        JITDUMP("  Cost with: " FMT_WT "\n", costWith);
+        JITDUMP("  Cost without: " FMT_WT "\n", costWithout);
+
+        if (costWith < costWithout)
+        {
+            JITDUMP("  Promoting replacement\n");
+            return true;
+        }
+
+#ifdef DEBUG
+        if (comp->compStressCompile(Compiler::STRESS_GENERALIZED_PROMOTION_COST, 25))
+        {
+            JITDUMP("  Promoting replacement due to stress\n");
+            return true;
+        }
+#endif
+
+        JITDUMP("  Disqualifying replacement\n");
+        return false;
+    }
+
+#ifdef DEBUG
+    void Dump(unsigned lclNum)
+    {
+        if (m_accesses.size() <= 0)
+        {
+            return;
+        }
+
+        printf("Accesses for V%02u\n", lclNum);
+        for (Access& access : m_accesses)
+        {
+            if (access.AccessType == TYP_STRUCT)
+            {
+                printf("  [%03u..%03u)\n", access.Offset, access.Offset + access.Layout->GetSize());
+            }
+            else
+            {
+                printf("  %s @ %03u\n", varTypeName(access.AccessType), access.Offset);
+            }
+
+            printf("    #:                             (%u, " FMT_WT ")\n", access.Count, access.CountWtd);
+            printf("    # assigned from:               (%u, " FMT_WT ")\n", access.CountAssignmentSource,
+                   access.CountAssignmentSourceWtd);
+            printf("    # assigned to:                 (%u, " FMT_WT ")\n", access.CountAssignmentDestination,
+                   access.CountAssignmentDestinationWtd);
+            printf("    # as call arg:                 (%u, " FMT_WT ")\n", access.CountCallArgs,
+                   access.CountCallArgsWtd);
+            printf("    # as implicit by-ref call arg: (%u, " FMT_WT ")\n", access.CountCallArgsByImplicitRef,
+                   access.CountCallArgsByImplicitRefWtd);
+            printf("    # as on-stack call arg:        (%u, " FMT_WT ")\n", access.CountCallArgsOnStack,
+                   access.CountCallArgsOnStackWtd);
+            printf("    # as retbuf:                   (%u, " FMT_WT ")\n", access.CountPassedAsRetbuf,
+                   access.CountPassedAsRetbufWtd);
+            printf("    # as returned value:           (%u, " FMT_WT ")\n\n", access.CountReturns,
+                   access.CountReturnsWtd);
+        }
+    }
+#endif
+};
+
+// Visitor that records information about uses of struct locals.
+class LocalsUseVisitor : public GenTreeVisitor<LocalsUseVisitor>
```

Comment:
It should be possible to switch this to use the linked list of locals instead. I will look into that once I get to a point of working on throughput.